InteractiveRender : Add message log

themissingcow opened the pull request:
As part of #3419, we require a mechanism to collect renderer output, and feed this into a log viewer within the Gaffer UI.

This commit adds new cheap-to-copy data types to hold messages, along with improved widgets to display and search them in the Gaffer UI. 

![image](https://user-images.githubusercontent.com/896779/83901288-69a80f80-a752-11ea-8d98-d6e1d0820758.png)

It should be backwards compatible with existing code, aside from the removal of the deprecated `MessageWidget.appendMessage` method.

Note: The `MessageWidget` changes seem to have introduced an intermittent `Internal C++ object already deleted` error. I've not been able to reliably reproduce this to fix it yet. It maybe to do with the need to hold a strong reference to a `QAction` [here](https://github.com/GafferHQ/gaffer/commit/3a1868662633246b9bf606fca65a70ed3147a895#diff-025a5d68e8d612161cc57c7ca80cea1bR426).
```
Traceback (most recent call last):
  File "/Volumes/Data/dev/gaffer-build/python/GafferUI/Widget.py", line 928, in eventFilter
    raise e
RuntimeError: Internal C++ object (PySide2.QtWidgets.QWidget) already deleted.
```
I'll attempt to get this resolved during the course of this review.

@johnhaddon There are a few contentious parts in some of the ancillary commits so feel free to suggest alternatives, etc :)

themissingcow commented:
@johnhaddon I still also need to make the sleeper Op and double check the blocking UI thread message handling.

themissingcow commented:
@andrewkaufman (assuming I've fixed that linux compile error), it'd be great if you had any time to check how this works with the more extended integration at IE.

andrewkaufman commented:
Thanks I'll try to take a look, but likely not until mid next week. Hope that's ok.

themissingcow commented:
> Thanks I'll try to take a look, but likely not until mid next week. Hope that's ok.

Thanks. Yeah totally, sorry, know there is a lot going on there right now. Just whenever you can. I'm sure there will be many a revision to this to be made :)

johnhaddon commented on include/GafferScene/InteractiveRender.h:
```diff
@@ -37,11 +37,12 @@
 #ifndef GAFFERSCENE_INTERACTIVERENDER_H
 #define GAFFERSCENE_INTERACTIVERENDER_H
 
+#include "Gaffer/TypedObjectPlug.h"
```
Move to Gaffer section below GafferScene.

themissingcow replied:
https://github.com/GafferHQ/gaffer/pull/3790/commits/04ec13379be81f71ca9bdf34063e38ef9861a24e

johnhaddon commented on include/GafferScene/InteractiveRender.h:
```diff
@@ -83,13 +84,20 @@ class GAFFERSCENE_API InteractiveRender : public Gaffer::Node
 
+		virtual void affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const override;
+
+		IECore::MessageHandler *messageHandler() const;
```
Perhaps it'll become obvious why during the rest of the review, but I wasn't expecting this to be exposed. I thought the `messagesPlug()` was the public interface?

themissingcow replied:
Ahh yeah, it was so we could have an integration test for the message handling overhead in `InteractiveRender` without needing a mock renderer, namely https://github.com/GafferHQ/gaffer/pull/3790/commits/9a7b69143e80c4e4ab925678bcc7322ea95f38d4

themissingcow replied:
As per our chat  https://github.com/GafferHQ/gaffer/pull/3790/commits/d6517e5ef41dae2643a920256283997d06f822d9 and https://github.com/GafferHQ/gaffer/pull/3790/commits/b2220b3d56bfc3103f0a121739127fcc7ec9de3b

johnhaddon commented on python/GafferUI/MessageWidget.py:
```diff
@@ -246,8 +232,1152 @@ def handle( self, level, context, msg ) :
+		else :
+			column = _MessageTableModel.Column.Message
+
+		return self.sourceModel().index( proxyIndex.row() // 2, int(column) )
```
`//` Yes! Python 3 readiness badge achieved!

johnhaddon commented on python/GafferUI/MessageWidget.py:
```diff
@@ -246,8 +232,1152 @@ def handle( self, level, context, msg ) :
+
+		return self.sourceModel().index( proxyIndex.row() // 2, int(column) )
+
+	def mapSelectionFromSource( self, sourceSelection ) :
+
+		selection = QtCore.QItemSelection()
+		for i in sourceSelection.indexes() :
+			index = self.mapFromSource( i )
+			selection.select( index, index )
+		return selection
```
I'm probably missing something because I'm reading about proxy models while reviewing this, but how is this different from the default implementation provided by the base class? And if we need to override `mapSelectionFromSource`, why don't we need to override `mapSelectionToSource()`?

themissingcow replied:
Good spot, thanks. These were a hang over from 'when nothing was working'. https://github.com/GafferHQ/gaffer/pull/3790/commits/aa07f1f32c4299cdc55ec98691f3a7d27cc06be7

johnhaddon commented on python/GafferUI/MessageWidget.py:
```diff
@@ -246,8 +232,1152 @@ def handle( self, level, context, msg ) :
+			selection.select( index, index )
+		return selection
+
+	def flags( self, index ) :
+
+		return self.sourceModel().flags( self.mapToSource( index ) )
```
This one definitely seems to be the same as the default implementation. Can we simply omit it?

themissingcow replied:
https://github.com/GafferHQ/gaffer/pull/3790/commits/aa07f1f32c4299cdc55ec98691f3a7d27cc06be7

johnhaddon commented on python/GafferUI/MessageWidget.py:
```diff
@@ -246,8 +232,1152 @@ def handle( self, level, context, msg ) :
+		if parent.isValid() :
+			return 0
+
+		return len( _MessageTableModel.Column.values() )
```
This is showing up as a bit of a hotspot in VTune. Seems like it makes and throws away a tuple every time. If we stick with a Python implementation, might be worth using a constant here instead.

themissingcow replied:
https://github.com/GafferHQ/gaffer/pull/3790/commits/3749bef030b027cf04a097b9e7f04f778a0e16c2

johnhaddon commented on python/GafferDispatchUI/LocalDispatcherUI.py:
```diff
@@ -240,7 +240,8 @@ def __init__( self, jobPool, **kw ) :
 
 					with GafferUI.ListContainer( GafferUI.ListContainer.Orientation.Vertical, spacing=10, borderWidth=10, parenting = { "label"  : "Messages" } ) as self.__messagesTab :
-						self.__messageWidget = GafferUI.MessageWidget()
+						self.__messageWidget = GafferUI.MessageWidget( toolbars = True, follow = True )
```
Nice! As a follow up, it would be great to look at `_LocalJobsWindow.__updateMessages()`. It's currently doing some crazy clear-and-resend-all-the-messages thing. We can talk about that after though...

johnhaddon commented on python/GafferSceneUI/InteractiveRenderUI.py:
```diff
@@ -229,6 +229,45 @@ def __startPauseClicked( self, button ) :
+
+class _MessagesPlugValueWidget( GafferUI.PlugValueWidget ) :
+
+	def __init__( self, plug, *args, **kwargs) :
```
Remove unused `args`. Forward `kw` to `PlugValueWidget.__init__()`.

themissingcow replied:
https://github.com/GafferHQ/gaffer/pull/3790/commits/efe1f7546742d13fd6be2668c1b965578758fad2

johnhaddon commented on python/GafferSceneUI/InteractiveRenderUI.py:
```diff
@@ -229,6 +229,45 @@ def __startPauseClicked( self, button ) :
+
+		GafferUI.PlugValueWidget.__init__( self, self.__messages, plug )
+
+		self.visibilityChangedSignal().connect( Gaffer.WeakMethod( self._updateFromPlug ), scoped = False )
```
How important is the dont-update-when-invisible optimisation? Rather than implement it manually, perhaps we should use `@LazyMethod`, both for simplicity and also to better communicate intent?

themissingcow replied:
So much better, I hadn't realised how neat `@LazyMethod` was :) https://github.com/GafferHQ/gaffer/pull/3790/commits/9317c15f9b01b9b59352d415896ab03f753607de

themissingcow commented on src/Gaffer/IECorePreview/Messages.cpp:
```diff
@@ -0,0 +1,125 @@
+{
+	m_nextBucket.clear();
+	m_buckets.clear();
+	std::fill_n( m_counts, int(IECore::MessageHandler::Level::Invalid), 0 );
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/3c6700dc711e9248af7334127163450da066109f

themissingcow commented on include/Gaffer/Private/IECorePreview/MessagesData.h:
```diff
@@ -0,0 +1,58 @@
+namespace IECore
+{
+
+	IECORE_DECLARE_TYPEDDATA( MessagesData, IECorePreview::Messages, void, IECore::SharedDataHolder );
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/1a1230bf9d397057db1da8aba298373537ac920d

themissingcow commented on src/GafferModule/MessagesBinding.cpp:
```diff
@@ -0,0 +1,106 @@
+
+void GafferModule::bindMessages()
+{
+	class_<Message>( "Message", no_init )
```
Thats much better. https://github.com/GafferHQ/gaffer/pull/3790/commits/1cf6ae9894472706c76a62d3fcad6467f3ca7f8a

themissingcow commented on src/GafferModule/MessagesBinding.cpp:
```diff
@@ -0,0 +1,106 @@
+
+Message getItem( Messages &m, long index )
+{
+	if( index < 0 || index > long( m.size() ) )
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/af4e1219cd58e200ee77c40341b0c4381b1c5649

themissingcow commented on python/GafferTest/IECorePreviewTest/MessagesTest.py:
```diff
@@ -0,0 +1,265 @@
+		self.assertTrue( m1 != m3 )
+
+		differentMsgs = [ m2, m3, m4, m5 ]
+		self.assertEqual( len(set(differentMsgs)), len(differentMsgs) )
```
Good idea. https://github.com/GafferHQ/gaffer/pull/3790/commits/42a7208bf77f89ffa0598d1270c08defb0a45e27

themissingcow commented on src/GafferModule/MessagesBinding.cpp:
```diff
@@ -0,0 +1,106 @@
+		)
+		.def_readonly( "level", &Message::level )
+		.def_readonly( "context", &Message::context )
+		.def_readonly( "message", &Message::message )
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/af4e1219cd58e200ee77c40341b0c4381b1c5649

johnhaddon replied:
c218b1aa27517981b6d0aba0844b64fe2efdffad

themissingcow commented on src/GafferModule/MessagesBinding.cpp:
```diff
@@ -0,0 +1,106 @@
+		.def_readonly( "context", &Message::context )
+		.def_readonly( "message", &Message::message )
+		.def( "hash", &Message::hash )
+		.def( "__eq__", &Message::operator== )
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/cffb48a858452c6c558ef31c6de1a0b17c1177c9

themissingcow commented on src/GafferModule/MessagesBinding.h:
```diff
@@ -0,0 +1,47 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2011, John Haddon. All rights reserved.
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/5ecc3cfb0fea33e41ba4d3b538bf5937c81a5a13

themissingcow commented on python/GafferTest/IECorePreviewTest/MessagesTest.py:
```diff
@@ -0,0 +1,265 @@
+			self.assertEqual( m.count( l ), 0 )
+
+		m = Gaffer.Messages()
+		self.assertEqual( m.size(), 0 )
```
It is! https://github.com/GafferHQ/gaffer/pull/3790/commits/bf8a3c8f8e8be191e19bc6dcf1d9cbf482707e09

themissingcow commented on python/GafferTest/IECorePreviewTest/MessagesTest.py:
```diff
@@ -0,0 +1,265 @@
+			m1.add( Gaffer.Message( IECore.MessageHandler.Level( i % 5 ), "testMessagesCopy", str(i)  ) )
+
+		m2 = m1
+		self.assertIs( m2, m1 )
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/bf8a3c8f8e8be191e19bc6dcf1d9cbf482707e09

themissingcow commented on python/GafferTest/IECorePreviewTest/MessagesTest.py:
```diff
@@ -0,0 +1,265 @@
+
+		mm1 = [ m1[i] for i in range( m1.size() ) ]
+		mm3 = [ m3[i] for i in range( m3.size() ) ]
+		self.assertListEqual( mm1, mm3 )
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/bf8a3c8f8e8be191e19bc6dcf1d9cbf482707e09

themissingcow commented on python/GafferTest/IECorePreviewTest/MessagesTest.py:
```diff
@@ -0,0 +1,265 @@
+			m2.add( msg )
+
+		self.assertEqual( m1, m2 )
+		self.assertTrue( m1 == m2 )
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/bf8a3c8f8e8be191e19bc6dcf1d9cbf482707e09

themissingcow commented on python/GafferTest/IECorePreviewTest/MessagesTest.py:
```diff
@@ -0,0 +1,265 @@
+		self.assertFalse( m1 != m2 )
+
+		m3 = m1
+		self.assertIs( m3, m1 )
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/bf8a3c8f8e8be191e19bc6dcf1d9cbf482707e09

themissingcow commented on python/GafferTest/IECorePreviewTest/MessagesTest.py:
```diff
@@ -0,0 +1,265 @@
+	def testMessagesHash( self ) :
+
+		m1 = Gaffer.Messages()
+		h = m1.hash();
```
😃   https://github.com/GafferHQ/gaffer/pull/3790/commits/bf8a3c8f8e8be191e19bc6dcf1d9cbf482707e09

themissingcow commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -257,3 +367,105 @@ void InteractiveRender::stop()
+}
+
+void InteractiveRender::hash( const Gaffer::ValuePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/a5663ad5ce94a84067c9a2053d4f1926ec613816

themissingcow commented on src/GafferScene/InteractiveRender.cpp:
```diff
@@ -257,3 +367,105 @@ void InteractiveRender::stop()
+	if( output == messagesPlug() )
+	{
+		static_cast<ObjectPlug *>( output )->setValue( m_messageHandler->messages() );
+	}
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/a5663ad5ce94a84067c9a2053d4f1926ec613816

themissingcow commented on include/GafferSceneTest/InteractiveRender.h:
```diff
@@ -0,0 +1,56 @@
+
+}
+
+} // namespace GafferSceneTest
```
As per our chat  https://github.com/GafferHQ/gaffer/pull/3790/commits/d6517e5ef41dae2643a920256283997d06f822d9 and https://github.com/GafferHQ/gaffer/pull/3790/commits/b2220b3d56bfc3103f0a121739127fcc7ec9de3b

themissingcow commented on src/GafferSceneTest/InteractiveRender.cpp:
```diff
@@ -0,0 +1,98 @@
+// of the message container. It repeatedly pulls on the messages plug for
+// every message received. It isn't representative of real-world UI loads as it
+// runs in a single thread, but surfaces any efficiency changes in the
+// implementation of the messages container or within IteractiveRender.
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/d6517e5ef41dae2643a920256283997d06f822d9

themissingcow commented on src/GafferSceneTest/InteractiveRender.cpp:
```diff
@@ -0,0 +1,98 @@
+	ConstMessagesDataPtr outMsg;
+	size_t numUpdates = 0;
+
+	const auto dirtyConnection = node->plugDirtiedSignal().connect( [node, &outMsg, &numUpdates]( Gaffer::Plug *plug ) {
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/d6517e5ef41dae2643a920256283997d06f822d9

themissingcow commented on src/GafferSceneTest/InteractiveRender.cpp:
```diff
@@ -0,0 +1,98 @@
+// of the message container. It repeatedly pulls on the messages plug for
+// every message received. It isn't representative of real-world UI loads as it
+// runs in a single thread, but surfaces any efficiency changes in the
+// implementation of the messages container or within IteractiveRender.
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/d6517e5ef41dae2643a920256283997d06f822d9

themissingcow commented on src/GafferSceneTest/InteractiveRender.cpp:
```diff
@@ -0,0 +1,98 @@
+		}
+	} );
+
+	MessageHandler *handler = node->messageHandler();
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/d6517e5ef41dae2643a920256283997d06f822d9

themissingcow commented on python/GafferUI/Image.py:
```diff
@@ -154,6 +154,14 @@ def _qtPixmapDisabled( self ) :
+
+		icon = QtGui.QIcon( self._qtPixmapHighlighted() if highlighted else self._qtPixmap() )
+		icon.addPixmap( self._qtPixmapDisabled(), QtGui.QIcon.Disabled )
+		return icon
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/be4b533ace998908039f89a25577ef1c34ea3c5e

themissingcow commented on python/GafferUI/Button.py:
```diff
@@ -123,6 +125,16 @@ def getHasFrame( self ) :
+			self.setHighlighted( False )
+			self.__highlightedDueToHover = False
+
+		GafferUI.Widget.setEnabled( self, enabled )
```
Much better idea, thanks https://github.com/GafferHQ/gaffer/pull/3790/commits/bf242e0e8a58bf85ff65e944959b05d1b9b9a262

themissingcow commented on python/GafferUI/_StyleSheet.py:
```diff
@@ -1140,6 +1145,20 @@ def styleColor( key ) :
 	}
 
+	*[gafferClass="GafferUI.MessageWidget._MessageTableView"] {
+		font-family: "SFMono-Regular", "Consolas", "Liberation Mono", "Menlo", monospace;
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/0ff6f9f0332c9df4b6b31f9faa1cdcdf18c0b791

themissingcow commented on python/GafferUI/MessageWidget.py:
```diff
@@ -47,45 +47,93 @@
 class MessageWidget( GafferUI.Widget ) :
 
-	def __init__( self, **kw ) :
+	# Messages : For presenting a longer messages in detail. They are shown as line-wrapped paragraphs.
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/e5654258f8ae77b08d9b2b996b98a1ed35384ce0

themissingcow commented on python/GafferUI/MessageWidget.py:
```diff
@@ -246,8 +232,1152 @@ def handle( self, level, context, msg ) :
+				# thread may be queuing up loads of messages using self.messageHandler(), and those
+				# will get processed by processEvents(), resulting in a recursive call to appendMessage().
+				# If the other thread supplies messages fast enough and we don't guard against recursion
+				# then we can end up exceeding Python's stack limit very quickly.
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/e5654258f8ae77b08d9b2b996b98a1ed35384ce0

themissingcow commented on python/GafferUI/MessageWidget.py:
```diff
@@ -47,45 +47,93 @@
+
+			if toolbars :
+
+				self.__upperToolbar = GafferUI.ListContainer( GafferUI.ListContainer.Orientation.Horizontal, spacing = 4 )
```
Good catch, thanks. https://github.com/GafferHQ/gaffer/pull/3790/commits/e81e7fee21d600027977a921fce6dcbc2f3d2e93

themissingcow commented on python/GafferUI/MessageWidget.py:
```diff
@@ -47,45 +47,93 @@
+			if toolbars :
+
+				self.__lowerToolbar = GafferUI.ListContainer( GafferUI.ListContainer.Orientation.Horizontal )
+				with self.__lowerToolbar :
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/e81e7fee21d600027977a921fce6dcbc2f3d2e93

themissingcow commented on python/GafferUI/MessageWidget.py:
```diff
@@ -246,8 +232,1152 @@ def handle( self, level, context, msg ) :
+	def messagesChangedSignal( self ) :
+
+		if self.__messagesChangedSignal is None :
+			self.__messagesChangedSignal = GafferUI.WidgetSignal()
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/15c7dadf14b848973c6e437bec48e2b4fe0d2168

themissingcow commented on python/GafferUI/MessageWidget.py:
```diff
@@ -246,8 +232,1152 @@ def handle( self, level, context, msg ) :
+	# Emitted with the button pressed. The level can be accessed via button.level
+	def levelButtonClickedSignal( self ) :
+
+		return self.__levelButtonClickedSignal
```
I wanted to do this before, but had missed `Gaffer.Signal`'s, much better, thanks https://github.com/GafferHQ/gaffer/pull/3790/commits/bc1a69452a0c262509ffc03cadf7d7f59284d8fe

themissingcow commented on python/GafferUI/MessageWidget.py:
```diff
@@ -246,8 +232,1152 @@ def handle( self, level, context, msg ) :
+#                    independent of the current display.
+#
+# The table's view is driven from the __displayModel - which is set to either the __filterModel or __searchModel
+# depending on getetFocusSearchResults().
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/e5654258f8ae77b08d9b2b996b98a1ed35384ce0

themissingcow commented:
@johnhaddon I've added 100,000,000 fixup commits for you. If you are able to sign off on those, I'll squash them all in before starting on the performance tweaks.

themissingcow commented on python/GafferUI/_StyleSheet.py:
```diff
@@ -341,7 +341,7 @@ def styleColor( key ) :
 
 	QPlainTextEdit[gafferRole="Code"] {
-		font-family: monospace;
+		font-family: "SFMono-Regular", "Consolas", "Liberation Mono", "Menlo", monospace;
```
Sadly this seems to be broken now (Catalina):

```
Fixed Font: <PySide2.QtGui.QFont( ",-1,-1,5,50,0,0,0,0,0" )  at 0x11b300f80> family: ''
Variable Font: <PySide2.QtGui.QFont( ",-1,-1,5,50,0,0,0,0,0" )  at 0x11b300f80> family: ''
```

It seems to be returning the same font for `FixedFont` or `GeneralFont`, which has no family. This at least gives us explicit control...

themissingcow commented on src/GafferModule/MessagesBinding.cpp:
```diff
@@ -68,6 +70,22 @@ Message getItem( Messages &m, long index )
+
+}
+
+// TODO: It woudld be great if we could get a long out of MurmurHash without
+// needing to go via a string.
```
Sorry, my mistake - had it on my list to fix up this morning, then, well. didnt.

Not sure if this link will work, I made it changing the commit in the files changed tab, but it also seems to be the same as the link from the landing page 🤷‍♂️ 
https://github.com/GafferHQ/gaffer/pull/3790/commits/3f58a8bd664b4a957c4ccd014a4e944ef346c28b

themissingcow replied:
Lol, it seems to actually re-write the link for us!

The link in the comment text is:
`https://github.com/GafferHQ/gaffer/pull/3790/commits/3f58a8bd664b4a957c4ccd014a4e944ef346c28b`
But the parsed link url is:
`https://github.com/GafferHQ/gaffer/commit/3f58a8bd664b4a957c4ccd014a4e944ef346c28b`
🙃

themissingcow replied:
Does [this work](https://github.com/GafferHQ/gaffer/pull/3790/commits/3f58a8bd664b4a957c4ccd014a4e944ef346c28b) if we do it manually?

johnhaddon replied:
Yep, I just discovered that too while trying to provide you with a good link on my UI PR :(

johnhaddon replied:
> Does this work if we do it manually?

Aha! Yes it does! Bit of a pain though - given the infrequency of second comments, perhaps it's better to put the onus on the reviewer?

themissingcow replied:
https://github.com/GafferHQ/gaffer/pull/3790/commits/4e874f95a8bf62c116d322743850ef32f8b287f2#diff-0cad84d3112e7a9bf1dc32edc2102d72R50

themissingcow commented on python/GafferUI/MessageWidget.py:
```diff
@@ -240,12 +240,7 @@ def handle( self, level, context, msg ) :
-				# will get processed by processEvents(), resulting in a recursive call to appendMessage().
-				# If the other thread supplies messages fast enough and we don't guard against recursion
-				# then we can end up exceeding Python's stack limit very quickly.
+				# We want to update the ui as they occur, so we force the event loop to clear here.
```
https://github.com/GafferHQ/gaffer/pull/3790/commits/0e3093a02f6c091d882f0b0a46cf2dce92728c01#diff-025a5d68e8d612161cc57c7ca80cea1bR245-R247 help?

themissingcow replied:
Bah, another unhelpful link (it doesn't expand the section)

![image](https://user-images.githubusercontent.com/896779/84383767-ce48eb80-abe4-11ea-9002-73c969e4c2f2.png)

themissingcow commented:
@johnhaddon I've implemented the performance tweaks we spoke of, namely : 
 - Combine context/message columns via display transform, turn off column resizing.
 - Only dirty changed parts of the model (via `Messages.firstDifference`).
 - `@LazyMethod` for widget update.
 - Fixed and improved search result navigation performance.

Ill carry on with performance testing, and regression testing with `OpDialogue`.

themissingcow commented:
> I also get the "skip to the end" button greyed out even when the I'm not scrolled fully to the end. I can reproduce this reliably by scrolling all the way to the end, and then hitting the up arrow on the scrollbar. I have to hit it 5 times before the button becomes enabled again.

This had been by design, to make it easier to manually scroll back to the end with a tablet (where its easy to joggle by a couple of pixels):

https://github.com/GafferHQ/gaffer/blob/68eddf1fcfb90c618d438e2f13a0f3af1f0f0a34/python/GafferUI/MessageWidget.py#L998

Can get rid of that though if preferred.

themissingcow commented:
![image](https://user-images.githubusercontent.com/896779/84516368-9fef0d00-acc5-11ea-80dc-8592c6f4234c.png)
How much better is this than 50!

themissingcow commented:
> Can you confirm that you've tested with OpDialogue though, as I think this is still an important use case at IE.

Checked, and `OpDialogue` not running the op in the BG does update during execution if messages are logged.

themissingcow commented:
> Can get rid of that though if preferred.

@johnhaddon https://github.com/GafferHQ/gaffer/pull/3790/commits/24d9e75ea3719d1dda433a2cead7dca70eba0b75

themissingcow commented on python/GafferUI/MessageWidget.py:
```diff
@@ -1073,6 +1067,87 @@ def __plainTextForMessages( self, messageIndices ) :
+# Having context as a separate column consequently requires either a fixed width, specified
+# by the parent UI, or truncated contents. Neither or which are ideal. This proxy combines
+# context/message such that we don't have to worry about how long the context string is.
+# It would same some boilerplate if we derived from QIdentityProxyModel (though strictly, this
```
Corrected here https://github.com/GafferHQ/gaffer/blob/f15e0fdc5db87dd69ab5bfb7b1bd45c188edef16/python/GafferUI/MessageWidget.py#L1096

themissingcow commented on python/GafferUI/MessageWidget.py:
```diff
@@ -1073,6 +1067,87 @@ def __plainTextForMessages( self, messageIndices ) :
+		if sourceIndex.column() == 2 :
+			return self.index( sourceIndex.row(), 1 )
+		else :
+			return self.index( sourceIndex.row(), sourceIndex.column() )
```
Further to our offline chat, I think we need this to allow this to work properly:

https://github.com/GafferHQ/gaffer/blob/f15e0fdc5db87dd69ab5bfb7b1bd45c188edef16/python/GafferUI/MessageWidget.py#L1180

themissingcow commented on include/Gaffer/Private/IECorePreview/Messages.h:
```diff
@@ -91,6 +91,11 @@ class GAFFER_API Messages
 
+		/// The index of the first message that differs to the messages in
+		/// the other container. The value may be beyond the bounds of this
+		/// container if the other is larger or both are identical.
```
Think the boost version works well, good idea, thanks:

https://github.com/GafferHQ/gaffer/pull/3790/commits/20fd05160b2e301d25a97883b81d9eadbbacc548#diff-657600cc838586d4a69d2b9e83fd163fR96-R101

themissingcow replied:
Worth a sanity check of https://github.com/GafferHQ/gaffer/pull/3790/commits/20fd05160b2e301d25a97883b81d9eadbbacc548#diff-0cad84d3112e7a9bf1dc32edc2102d72R86-R88too

themissingcow commented on python/GafferUI/MessageWidget.py:
```diff
@@ -410,8 +411,10 @@ def __init__( self, tableView, **kw ) :
+			# Edited catches focus-out and makes sure we update the search text
 			self.__searchField.editingFinishedSignal().connect( Gaffer.WeakMethod( self.__textEdited ), scoped = False )
-			self.__searchField.activatedSignal().connect( Gaffer.WeakMethod( self.__textEdited ), scoped = False )
+			# Activated allows <enter> to jump to repeatedly jump to the next search result
```
Fixed in https://github.com/GafferHQ/gaffer/blob/f15e0fdc5db87dd69ab5bfb7b1bd45c188edef16/python/GafferUI/MessageWidget.py#L416

themissingcow commented on python/GafferUI/MessageWidget.py:
```diff
@@ -863,60 +861,82 @@ def __searchResultsChanged( self, *unused ) :
+		model = self.__searchModel
+		selectedDisplayRow = displayIndex.row()
+
+		# As bisect needs an sequence type, but we don't want to pre-geneate a list of all result
```
Fixed in https://github.com/GafferHQ/gaffer/blob/f15e0fdc5db87dd69ab5bfb7b1bd45c188edef16/python/GafferUI/MessageWidget.py#L927

themissingcow commented on src/GafferModule/MessagesBinding.cpp:
```diff
@@ -86,6 +86,11 @@ long hashMessages( const Messages &m )
 
+std::string reprMessagesData( const MessagesData *m )
+{
+	return std::string( "Gaffer.Private.IECorePreview.MessagesData()" );
```
Updated here https://github.com/GafferHQ/gaffer/pull/3790/commits/20fd05160b2e301d25a97883b81d9eadbbacc548#diff-0cad84d3112e7a9bf1dc32edc2102d72R98-R101

themissingcow commented on python/GafferSceneUI/InteractiveRenderUI.py:
```diff
@@ -150,7 +150,7 @@ def __imageIsRendering( self, imagePlug ) :
 			return False
 
-	@GafferUI.LazyMethod()
+	@GafferUI.LazyMethod( deferUntilPlaybackStops = True )
```
Updated https://github.com/GafferHQ/gaffer/pull/3790/commits/c6ae670c6efece1dbb8e5abbfc733f6c2df97545 to leave that alone for now...

themissingcow commented:
@johnhaddon Thanks for catching those, and the good ideas. As per our chat, I've squished all this back so hopefully we're a little closer...

johnhaddon commented on python/GafferUI/MessageWidget.py:
```diff
@@ -246,8 +234,1225 @@ def handle( self, level, context, msg ) :
+		# scrolls to the top, and then is re-scrolled back to the
+		# bottom.
+
+		firstDifference = messages.firstDifference( self.__messages ) if self.__messages else None
```
`self.__messages` is false when we have an empty messages container, in which case we would think we had no differences. Should be `if self.__messages is not None`?

themissingcow replied:
Good spot, thanks! I think between us we've managed to get the fact that that line was almost 100% wrong. The default should have been `0` too! I've fixed it in place:

https://github.com/GafferHQ/gaffer/blob/a9a75df3660b88bbfe4940e23969ecf5e59d8a52/python/GafferUI/MessageWidget.py#L1367

johnhaddon replied:
Teamwork!

johnhaddon commented:
LGTM - feel free to merge once CI is done.

themissingcow commented:
> LGTM - feel free to merge once CI is done.

Thanks for the 🦅 👀  on this one John!
